exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java

```diff
@@ -68,8 +78,37 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
     // Submit user query to Drillbit work queue.
     final QueryId queryId = workManager.getUserWorker().submitWork(webUserConnection, runQuery);
 
+    heapMemoryFailureThreshold = workManager.getContext().getConfig().getDouble( ExecConstants.HTTP_QUERY_FAIL_LOW_HEAP_THRESHOLD );
+    boolean isComplete = false;
+    boolean nearlyOutOfHeapSpace = false;
+    float usagePercent = getHeapUsage();
+
     // Wait until the query execution is complete or there is error submitting the query
-    webUserConnection.await();
+    logger.debug("Wait until the query execution is complete or there is error submitting the query");
+    do {
+      try {
+        isComplete = webUserConnection.await(TimeUnit.SECONDS.toMillis(1)); /*periodically timeout to check heap*/
+      } catch (Exception e) { }
+
+      usagePercent = getHeapUsage();
+      if (usagePercent >  heapMemoryFailureThreshold) {
+        nearlyOutOfHeapSpace = true;
+      }
+    } while (!isComplete && !nearlyOutOfHeapSpace);
+
+    //Fail if nearly out of heap space
+    if (nearlyOutOfHeapSpace) {
+      workManager.getBee().getForemanForQueryId(queryId)
+        .addToEventQueue(QueryState.FAILED,
+            UserException.resourceError(
+                new Throwable(
+                    "Query submitted through the Web interface was failed due to diminishing free heap memory ("+ Math.floor(((1-usagePercent)*100)) +"% free). "
```

Comment:
We could make this friendlier :)
"There is not enough heap memory to run this query using the web interface. Please try a query with fewer columns or with a filter or limit condition to limit the data returned. You can also try an ODBC/JDBC client"

Reply:
I thought it would be useful to show the level at which the free memory was prior to cancellation. I suspect that people will see a GC'ed Drillbit after the cancellation and wonder why Drill is complaining of no sufficient heap.  Hence, the wording.

Reply:
Oh yes. That would be useful information to add to the message. But since the error is triggered at 0.85, the number would always be 15% :)